salt/utils/openstack/nova.py

```diff
@@ -355,21 +365,14 @@ def _old_init(self, username, project_id, auth_url, region_name, password, os_au
 
     def _v3_setup(self, region_name):
         if region_name is not None:
-            servers_endpoints = get_entry(self.catalog, 'type', 'compute')['endpoints']
-            self.kwargs['bypass_url'] = get_entry_multi(
-                servers_endpoints,
-                [('region', region_name), ('interface', 'public')]
-            )['url']
+            self.kwargs['bypass_url'] = get_endpoint_url_v3(self.catalog, 'compute', region_name)
 
         self.compute_conn = client.Client(version=self.version, session=self.session, **self.client_kwargs)
 
         volume_endpoints = get_entry(self.catalog, 'type', 'volume', raise_error=False).get('endpoints', {})
```

Comment:
this line can be removed now, since you are passing the whole catalog.

Reply:
I think it is still useful when user does not specify a region.
In this case we proceed without invoking the new logic.

I just realized that here (unlike in _v2_setup) we set self.kwargs['bypass_url'],
but then use self.client_kwargs to construct both Nova and Cinder clients.
Is that intentional?

Reply:
Ahh you are right

also, these should be flipped over to specifying bypass_url in client_kwargs.

Most of this will be going away in Oxygen, as we will be recommending using the openstack driver I have on my list to get done by the feature freeze, and we will be using the shade library from openstack infra, until shade is merged into the openstacksdk.

Reply:
Ok, done. Not sure why it created an instance in the correct region before.
I also added logging to verify that correct bypass_urls are picked.